src/parsing_visualizer.cpp

```diff
@@ -0,0 +1,172 @@
+
+std::string expect_one_param(const expression &expr) {
+    if (expr.get_parameters().empty()) return "?";
+    else if (expr.get_parameters().size() == 1) {
+        return stringify(expr.get_parameters().front());
+    } else return stringify_op(expr, "??", "()");
+}
+
+std::string stringify(const expression& expr, bool stmt) {
+    switch (expr.get_type()) {
+        case expression_type::nop    :
+            return "";
+        case expression_type::string :
+            return "\\\"" + expr.get_str_value() + "\\\"";
+        case expression_type::number :
+            return std::to_string(expr.get_number_val());
+        case expression_type::identifier  : {
+            const char *ident_type[] = {"?", "Func", "Param", "Var", "Stmt"};
+            return ident_type[(int) expr.get_identifier().get_ident_type()] +
+                   std::to_string(expr.get_identifier().get_index_within()) + "\\\"" + expr.get_identifier().get_name() +
+                   "\\\"";
+        }
+        case expression_type::addition    :
+            return stringify_op(expr, " + ", "()");
+        case expression_type::multiplication:
+            return stringify_op(expr, " * ", "()");
+        case expression_type::division:
+            return stringify_op(expr, " / ", "()");
+        case expression_type::equality     :
+            return stringify_op(expr, " == ", "()");
+        case expression_type::compare_and   :
+            return stringify_op(expr, " && ", "()");
+        case expression_type::expression_sequence  :
+            return stmt ? stringify_op(expr, "; ", "{}", true) : stringify_op(expr, ", ", "()");
+        case expression_type::negation    :
+            return "-(" + expect_one_param(expr) + ")";
+        case expression_type::copy   :
+            return "(" + stringify(expr.get_parameters().back()) + " = " + stringify(expr.get_parameters().front()) + ")";
+        case expression_type::function_call  :
+            return "(" + (expr.get_parameters().empty() ? "?" : stringify(expr.get_parameters().front())) + ")" +
+                   stringify_op(expr, ", ", "()", false, 1);
+        case expression_type::compare_loop   :
+            return "while " + stringify(expr.get_parameters().front()) + " " + stringify_op(expr, "; ", "{}", true, 1);
+        case expression_type::retrn    :
+            return "return " + expect_one_param(expr);
+    }
+    return "?";
+}
+
+std::string stringify(const function &f) {
+    return stringify(f.code, true);
+}
+
+std::string get_next_node_token(const std::string &string, unsigned &start_idx) {
+    if (start_idx >= string.size()) {
+        return "";
+    }
+
+    // Loop till non-space character appear
+    for (unsigned i = start_idx; i < string.size(); ++i) {
+        if (string[i] == ' ' || string[i] == ';') {
+            continue;
+        }
+        start_idx = i;
+        break;
+    }
+
+    unsigned end_idx;
+    for (end_idx = start_idx; end_idx < string.size(); ++end_idx) {
```

Comment:
Prefer modern style loops: `for (const auto &el: collection) {}`

Reply:
Generally said, yes, but for this function I return modern ```substr``` in the end and it works with a number and therefore it's easier to work with indexes in this exact situation. Glad to here more if you have a better solution :smiley: